### ISSUE TITLE
firefox gets wrong timezone on NixOS

### ISSUE THREAD

**Issue by sg-qwt** [NixOS/nixpkgs#238025]
### Steps To Reproduce
1. NixOS has correctly set timezone with  `time.timeZone = "Asia/Shanghai"`

2. run `date` and `timedatectl` all return correct timezone
3. open Chromium and test timezone with https://webbrowsertools.com/timezone/ timezone is correct
4. open Firefox and test timezone with https://webbrowsertools.com/timezone/ timezone is wrong, in my case it shows `GMT-0500 (Central Daylight Time)`
(doesn't matter if it's firefox or firefox-beta, `privacy.resistFingerprinting` is false)
5. run firefox with TZ set works, eg `TZ=Asia/Shanghai /nix/store/pnm2qqfwpg7gqvybp2p0cghci3nm94my-firefox-114.0b7/bin/firefox`, visit https://webbrowsertools.com/timezone/ and timezone is correct now
6. my observation is Firefox needs TZ explicit set, which NixOS doesn't

### Additional context
Possibly  related to https://github.com/NixOS/nixpkgs/issues/23053

**Comment by ralismark** [comment]
Had the same issue, I noticed it from discord and facebook showing times in UTC. Setting `TZ=:/etc/localtime` also fixed this.

I suspect this is a recent (last few weeks) issue, since I feel like it was fine before.

**Comment by vcunat** [comment]
I can't reproduce the issue in `firefox` on current nixpkgs master (f891c1c65e), at least when running NixOS 23.05.  (`$TZ` is not set when I start up shell.)

**Comment by kirillrdy** [comment]
I can reproduce using that commit 

```
git checkout f891c1c65e31a7ebebb57cbedb3a63d0012146c1
nix-build -A firefox
./result/bin/firefox https://webbrowsertools.com/timezone
```


![image](https://github.com/NixOS/nixpkgs/assets/12160/aa412503-dfde-4193-a5ef-5e44367ac68d)

**Comment by kirillrdy** [comment]
works if sandbox is disabled 

```
MOZ_DISABLE_CONTENT_SANDBOX=1 ./result/bin/firefox https://webbrowsertools.com/timezone/
```

**Comment by MobSenpai** [comment]
I need to use MOZ_DISABLE_RDD_SANDBOX = "1" for va-api to work

**Comment by MobSenpai** [comment]
> works if sandbox is disabled
> 
> ```
> MOZ_DISABLE_CONTENT_SANDBOX=1 ./result/bin/firefox https://webbrowsertools.com/timezone/
> ```
I need to use MOZ_DISABLE_RDD_SANDBOX = "1" for va-api to work

**Comment by MobSenpai** [comment]
Set the TZ = "${config.time.timeZone}"; environment variable

**Comment by MobSenpai** [comment]
> Set the `TZ = "${config.time.timeZone}";` environment variable

I can just manually set this but knowing that It worked in the previous update my mind just don't allow me to add nother line to my config, so for now until they add it in the browser or soemthing I will just launch firefox using `TZ=mytimezone firefox`

**Comment by kirillrdy** [comment]
with firefox 115 timezone issue is back ... I guess until we hear back from upstream workaround is to set TZ

**Comment by kirillrdy** [comment]
> pinned

firefox 115 requires icu73, so it's either `TZ=":/etc/localtime"` or wait for upstream

**Comment by S-NA** [comment]
If it is a sandbox issue, would not using `security.sandbox.content.read_path_whitelist` in `about:config` be a workaround to this? I cannot reproduce the issue for some reason to test.

**Comment by kirillrdy** [comment]
> If it is a sandbox issue, would not using `security.sandbox.content.read_path_whitelist` in `about:config` be a workaround to this? I cannot reproduce the issue for some reason to test.

for `read_path_whitelist` ? can we set it programmatically ?

as for reproducing the issue, I have `nixos-rebuild build-vm --flake github:kirillrdy/dotfiles/firefox#tsutenkaku` that you can try if you wish

setting `MOZ_DISABLE_CONTENT_SANDBOX=1` fixes the issue, hence why I am confident its sandbox related

but whats strange is 

1) there are a few people that can not reproduce the issue
2) when i run with MOZ_SANDBOX_LOGGING=1
I get 
```
Sandbox: Recording mapping /nix/store/lyw3dwzv3wpvklmm5ky3zw3zxvx3nmvs-tzdata-2023c/share/zoneinfo/Australia/Melbourne -> /etc/localtime
Sandbox: Recording mapping /etc/zoneinfo/Australia/Melbourne -> /etc/localtime
Sandbox: Recording mapping /nix/store/lyw3dwzv3wpvklmm5ky3zw3zxvx3nmvs-tzdata-2023c/share/zoneinfo/Australia/Melbourne -> /etc/localtime
```

so maybe some sort of race condition, hopefully upstream can shine some light on the issue

**Comment by ralismark** [comment]
> If it is a sandbox issue, would not using `security.sandbox.content.read_path_whitelist` in `about:config` be a workaround to this? I cannot reproduce the issue for some reason to test.

setting `security.sandbox.content.read_path_whitelist` to `/etc/localtime,/etc/static/zoneinfo,/etc/static/zoneinfo/,/etc/zoneinfo,/etc/zoneinfo/,/nix/store/jhgh02lyizd1kyl71brvc01ygsmgi40a-tzdata-2023c/share/zoneinfo/,/nix/store/jhgh02lyizd1kyl71brvc01ygsmgi40a-tzdata-2023c/share/zoneinfo,/nix/store/9sv16r56wlhx39ms050ksi2axc1i5zn2-etc/etc,/nix/store/9sv16r56wlhx39ms050ksi2axc1i5zn2-etc/etc/` (basically, every single path that should be relevant) still shows the wrong timezone on https://webbrowsertools.com/timezone/

**Comment by mateusvmv** [comment]
Workaround, setting the TZ environment variable:
```nix
{ pkgs, ... }:
let
        patched-firefox = pkgs.symlinkJoin {
                name= "patched-firefox";
                paths = [ pkgs.firefox ];
                buildInputs = [ pkgs.makeWrapper ];
                postBuild = ''
                        wrapProgram "$out/bin/firefox" --set TZ /etc/localtime
                '';
        };
in
{
        home.packages = with pkgs; [
                patched-firefox
        ];
}
```

**Comment by MobSenpai** [comment]
Wth it's still not done yet?

**Comment by mweinelt** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?id=1848615

wontfix in 116, affected for 117/118, so an upstream fix will likely arrive in a 117 point release. Firefox 117.0 is due early next week.

**Comment by kirillrdy** [comment]
looks like it will be fixed in 119

**Comment by mweinelt** [comment]
Fix targeted for 117.0.1.
